File: assignments/haskell/crypto-square/crypto-square_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import System.Exit (ExitCode(..), exitWith)
import CryptoSquare (normalizePlaintext,
                     squareSize,
                     plaintextSegments,
                     ciphertext,
                     normalizeCiphertext)

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly $ runTestTT $ TestList
       [ TestLabel "normalizePlaintext" $ TestList normalizePlaintextTests
       , TestLabel "squareSize" $ TestList squareSizeTests
       , TestLabel "plaintextSegments" $ TestList plaintextSegmentsTests
       , TestLabel "ciphertext" $ TestList ciphertextTests
       , TestLabel "normalizeCiphertext" $ TestList normalizeCiphertextTests
       ]

normalizePlaintextTests :: [Test]
normalizePlaintextTests = map TestCase
  [ "splunk" @=? normalizePlaintext "s#!@$%plunk"
  , "123go" @=? normalizePlaintext "1, 2, 3 GO!"
  ]

squareSizeTests :: [Test]
squareSizeTests = map TestCase
  [ 2 @=? squareSize "1234"
  , 3 @=? squareSize "123456789"
  , 4 @=? squareSize "123456789abc" ]

plaintextSegmentsTests :: [Test]
plaintextSegmentsTests = map TestCase
  [ ["neverv", "exthin", "eheart", "withid", "lewoes"] @=?
    plaintextSegments "Never vex thine heart with idle woes."
  , ["zomg", "zomb", "ies"] @=?
    plaintextSegments "ZOMG! ZOMBIES!!!"
  ]

ciphertextTests :: [Test]
ciphertextTests = map TestCase
  [ "tasneyinicdsmiohooelntuillibsuuml" @=?
    ciphertext "Time is an illusion. Lunchtime doubly so."
  , "wneiaweoreneawssciliprerlneoidktcms" @=?
    ciphertext "We all know interspecies romance is weird."
  ]

normalizeCiphertextTests :: [Test]
normalizeCiphertextTests = map TestCase
  [ "msemo aanin dninn dlaet ltshu i" @=?
    normalizeCiphertext "Madness, and then illumination."
  , "vrela epems etpao oirpo" @=?
     normalizeCiphertext "Vampires are people too!"
  ]

File: assignments/clojure/crypto-square/crypto_square_test.clj
(ns crypt_square.test (:require [clojure.test :refer :all]))
(load-file "crypto_square.clj")

(deftest normalize-splunk
  (is (= "splunk" (normalize-plaintext "s#!@$%plunk"))))
(deftest normalize-with-punctuation
  (is (= "123go" (normalize-plaintext "1, 2, 3 GO!"))))

(deftest square-2
  (is (= 2 (square-size "1234"))))
(deftest square-3
  (is (= 3 (square-size "123456789"))))
(deftest square-4
  (is (= 4 (square-size "123456789abc"))))

(deftest segments
  (is (= ["neverv", "exthin", "eheart", "withid", "lewoes"]
         (plaintext-segments "Never vex thine heart with idle woes."))))
(deftest segments-2
  (is (= ["zomg", "zomb", "ies"]
         (plaintext-segments "ZOMG! ZOMBIES!!!"))))

(deftest cipher-1
  (is (= "tasneyinicdsmiohooelntuillibsuuml"
         (ciphertext "Time is an illusion. Lunchtime doubly so."))))
(deftest cipher-2
  (is (= "wneiaweoreneawssciliprerlneoidktcms"
         (ciphertext "We all know interspecies romance is weird."))))
(deftest cipher-3
  (is (= "msemo aanin dninn dlaet ltshu i"
         (normalize-ciphertext "Madness, and then illumination."))))
(deftest cipher-4
  (is (= "vrela epems etpao oirpo"
         (normalize-ciphertext "Vampires are people too!"))))

(run-tests)

